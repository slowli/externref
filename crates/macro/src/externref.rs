use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{
    parse::Error as SynError, punctuated::Punctuated, spanned::Spanned, Attribute, Expr, ExprLit,
    FnArg, ForeignItem, GenericArgument, Ident, ItemFn, ItemForeignMod, Lit, LitStr, Meta, PatType,
    Path, PathArguments, Signature, Token, Type, TypePath, Visibility,
};

use std::{collections::HashMap, mem};

use crate::ExternrefAttrs;

fn check_abi(
    target_name: &str,
    abi_name: Option<&LitStr>,
    root_span: &impl ToTokens,
) -> Result<(), SynError> {
    let abi_name = abi_name.ok_or_else(|| {
        let msg = format!("{target_name} must be marked with `extern \"C\"`");
        SynError::new_spanned(root_span, msg)
    })?;
    if abi_name.value() != "C" {
        let msg = format!(
            "Unexpected ABI {} for {target_name}; expected `C`",
            abi_name.value()
        );
        return Err(SynError::new(abi_name.span(), msg));
    }
    Ok(())
}

fn attr_expr(attrs: &[Attribute], name: &str) -> Result<Option<Expr>, SynError> {
    let attr = attrs.iter().find(|attr| attr.path().is_ident(name));
    let Some(attr) = attr else {
        return Ok(None);
    };

    let name_value = attr.meta.require_name_value()?;
    Ok(Some(name_value.value.clone()))
}

#[derive(Debug, Clone, Copy, PartialEq)]
enum SimpleResourceKind {
    Owned,
    Ref,
    MutRef,
}

impl SimpleResourceKind {
    fn is_resource(ty: &TypePath) -> bool {
        ty.path.segments.last().map_or(false, |segment| {
            segment.ident == "Resource"
                && matches!(
                    &segment.arguments,
                    PathArguments::AngleBracketed(args) if args.args.len() == 1
                )
        })
    }

    fn from_type(ty: &Type) -> Option<Self> {
        match ty {
            Type::Path(path) if Self::is_resource(path) => Some(Self::Owned),
            Type::Reference(reference) => {
                if let Type::Path(path) = reference.elem.as_ref() {
                    if Self::is_resource(path) {
                        return Some(if reference.mutability.is_some() {
                            Self::MutRef
                        } else {
                            Self::Ref
                        });
                    }
                }
                None
            }
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
enum ResourceKind {
    Simple(SimpleResourceKind),
    Option(SimpleResourceKind),
}

impl From<SimpleResourceKind> for ResourceKind {
    fn from(simple: SimpleResourceKind) -> Self {
        Self::Simple(simple)
    }
}

impl ResourceKind {
    fn parse_option(ty: &TypePath) -> Option<&Type> {
        let segment = ty.path.segments.last()?;
        if segment.ident == "Option" {
            if let PathArguments::AngleBracketed(args) = &segment.arguments {
                if args.args.len() == 1 {
                    if let GenericArgument::Type(ty) = args.args.first().unwrap() {
                        return Some(ty);
                    }
                }
            }
        }
        None
    }

    fn from_type(ty: &Type) -> Option<Self> {
        if let Some(kind) = SimpleResourceKind::from_type(ty) {
            return Some(kind.into());
        }

        if let Type::Path(path) = ty {
            Self::parse_option(path)
                .and_then(|inner_ty| SimpleResourceKind::from_type(inner_ty).map(Self::Option))
        } else {
            None
        }
    }

    fn simple_kind(self) -> SimpleResourceKind {
        match self {
            Self::Simple(simple) | Self::Option(simple) => simple,
        }
    }

    fn initialize_for_export(self, arg: &Ident, cr: &Path) -> TokenStream {
        let method_call = match self.simple_kind() {
            SimpleResourceKind::Owned => None,
            SimpleResourceKind::Ref => Some(quote!(.as_ref())),
            SimpleResourceKind::MutRef => Some(quote!(.as_mut())),
        };
        let unwrap = match self {
            Self::Option(_) => None,
            Self::Simple(_) => Some(quote!(.expect("null reference passed from host"))),
        };

        quote! {
            #cr::Resource::new(#arg) #method_call #unwrap
        }
    }

    fn prepare_for_import(self, arg: &Ident, cr: &Path) -> TokenStream {
        let arg = match self {
            Self::Simple(_) => quote!(core::option::Option::Some(#arg)),
            Self::Option(_) => quote!(#arg),
        };

        match self.simple_kind() {
            SimpleResourceKind::Ref | SimpleResourceKind::MutRef => {
                quote!(#cr::Resource::raw(#arg))
            }
            SimpleResourceKind::Owned => quote!(#cr::Resource::take_raw(#arg)),
        }
    }
}

#[derive(Debug, PartialEq)]
enum ReturnType {
    Default,
    NotResource,
    Resource(ResourceKind),
}

struct Function {
    name: Expr,
    arg_count: usize,
    resource_args: HashMap<usize, ResourceKind>,
    return_type: ReturnType,
    crate_path: Path,
}

impl Function {
    fn new(function: &ItemFn, attrs: &ExternrefAttrs) -> Result<Self, SynError> {
        let abi_name = function.sig.abi.as_ref().and_then(|abi| abi.name.as_ref());
        check_abi("exported function", abi_name, &function.sig)?;

        if let Some(variadic) = &function.sig.variadic {
            let msg = "Variadic functions are not supported";
            return Err(SynError::new_spanned(variadic, msg));
        }
        let export_name = attr_expr(&function.attrs, "export_name")?;
        Ok(Self::from_sig(&function.sig, export_name, attrs))
    }

    fn from_sig(sig: &Signature, name_override: Option<Expr>, attrs: &ExternrefAttrs) -> Self {
        let resource_args = sig.inputs.iter().enumerate().filter_map(|(i, arg)| {
            if let FnArg::Typed(PatType { ty, .. }) = arg {
                return ResourceKind::from_type(ty).map(|kind| (i, kind));
            }
            None
        });
        let return_type = match &sig.output {
            syn::ReturnType::Type(_, ty) => {
                ResourceKind::from_type(ty).map_or(ReturnType::NotResource, ReturnType::Resource)
            }
            syn::ReturnType::Default => ReturnType::Default,
        };
        let name = name_override.unwrap_or_else(|| {
            let str = sig.ident.to_string();
            syn::parse_quote!(#str)
        });

        Self {
            name,
            arg_count: sig.inputs.len(),
            resource_args: resource_args.collect(),
            return_type,
            crate_path: attrs.crate_path(),
        }
    }

    fn needs_declaring(&self) -> bool {
        !self.resource_args.is_empty() || matches!(self.return_type, ReturnType::Resource(_))
    }

    fn declare(&self, module_name: Option<&str>) -> impl ToTokens {
        let name = &self.name;
        let cr = &self.crate_path;
        let kind = if let Some(module_name) = module_name {
            quote!(#cr::FunctionKind::Import(#module_name))
        } else {
            quote!(#cr::FunctionKind::Export)
        };
        let externrefs = self.create_externrefs();

        quote! {
            #cr::declare_function!(#cr::Function {
                kind: #kind,
                name: #name,
                externrefs: #externrefs,
            });
        }
    }

    fn wrap_export(&self, raw: &ItemFn, export_name: Option<Attribute>) -> impl ToTokens {
        let cr = &self.crate_path;
        let export_name = export_name.unwrap_or_else(|| {
            let name = raw.sig.ident.to_string();
            syn::parse_quote!(#[export_name = #name])
        });
        let mut export_sig = raw.sig.clone();
        export_sig.abi = Some(syn::parse_quote!(extern "C"));
        export_sig.unsafety = Some(syn::parse_quote!(unsafe));
        export_sig.ident = Ident::new("__externref_export", export_sig.ident.span());

        let mut args = Vec::with_capacity(export_sig.inputs.len());
        for (i, arg) in export_sig.inputs.iter_mut().enumerate() {
            if let FnArg::Typed(typed_arg) = arg {
                let arg = Ident::new(&format!("__arg{i}"), typed_arg.pat.span());
                typed_arg.pat = Box::new(syn::parse_quote!(#arg));

                if let Some(kind) = self.resource_args.get(&i) {
                    typed_arg.ty = Box::new(syn::parse_quote!(#cr::ExternRef));
                    args.push(kind.initialize_for_export(&arg, cr));
                } else {
                    args.push(quote!(#arg));
                }
            }
        }

        let original_name = &raw.sig.ident;
        let delegation = quote!(#original_name(#(#args,)*));
        let delegation = match self.return_type {
            ReturnType::Resource(kind) => {
                export_sig.output = syn::parse_quote!(-> #cr::ExternRef);
                let output = Ident::new("__output", raw.sig.span());
                let conversion = kind.prepare_for_import(&output, cr);
                quote! {
                    let #output = #delegation;
                    #conversion
                }
            }
            ReturnType::NotResource => delegation,
            ReturnType::Default => quote!(#delegation;),
        };

        quote! {
            const _: () = {
                #export_name
                #export_sig {
                    #delegation
                }
            };
        }
    }

    fn wrap_import(&self, vis: &Visibility, mut sig: Signature) -> (TokenStream, Ident) {
        let cr = &self.crate_path;
        sig.unsafety = Some(syn::parse_quote!(unsafe));
        let new_ident = format!("__externref_{}", sig.ident);
        let new_ident = Ident::new(&new_ident, sig.ident.span());

        let mut args = Vec::with_capacity(sig.inputs.len());
        for (i, arg) in sig.inputs.iter_mut().enumerate() {
            if let FnArg::Typed(typed_arg) = arg {
                let arg = Ident::new(&format!("__arg{i}"), typed_arg.pat.span());
                typed_arg.pat = Box::new(syn::parse_quote!(#arg));

                if let Some(kind) = self.resource_args.get(&i) {
                    args.push(kind.prepare_for_import(&arg, cr));
                } else {
                    args.push(quote!(#arg));
                }
            }
        }

        let delegation = quote!(#new_ident(#(#args,)*));
        let delegation = match self.return_type {
            ReturnType::Resource(kind) => {
                let output = Ident::new("__output", sig.span());
                let init = kind.initialize_for_export(&output, cr);
                quote! {
                    let #output = #delegation;
                    #init
                }
            }
            ReturnType::NotResource => delegation,
            ReturnType::Default => quote!(#delegation;),
        };

        let wrapper = quote! {
            #[inline(never)]
            #vis #sig {
                unsafe { #cr::ExternRef::guard(); }
                #delegation
            }
        };
        (wrapper, new_ident)
    }

    fn create_externrefs(&self) -> impl ToTokens {
        let cr = &self.crate_path;
        let args_and_return_type_count = if matches!(self.return_type, ReturnType::Default) {
            self.arg_count
        } else {
            self.arg_count + 1
        };
        let bytes = (args_and_return_type_count + 7) / 8;

        let maybe_ret_idx = if matches!(self.return_type, ReturnType::Resource(_)) {
            Some(self.arg_count)
        } else {
            None
        };

        let set_bits = self.resource_args.keys().copied();
        #[cfg(test)] // sort keys in deterministic order for testing
        let set_bits = {
            let mut sorted: Vec<_> = set_bits.collect();
            sorted.sort_unstable();
            sorted.into_iter()
        };
        let set_bits = set_bits.chain(maybe_ret_idx);
        let set_bits = set_bits.map(|idx| quote!(.with_set_bit(#idx)));

        quote! {
            #cr::BitSlice::builder::<#bytes>(#args_and_return_type_count)
                #(#set_bits)*
                .build()
        }
    }
}

pub(crate) fn for_export(function: &mut ItemFn, attrs: &ExternrefAttrs) -> TokenStream {
    let parsed_function = match Function::new(function, attrs) {
        Ok(function) => function,
        Err(err) => return err.into_compile_error(),
    };
    let (declaration, export) = if parsed_function.needs_declaring() {
        // "Un-export" the function by removing the relevant attributes.
        function.sig.abi = None;
        let attr_idx = function.attrs.iter().enumerate().find_map(|(idx, attr)| {
            if attr.path().is_ident("export_name") {
                Some(idx)
            } else {
                None
            }
        });
        let export_name_attr = attr_idx.map(|idx| function.attrs.remove(idx));

        // Remove `#[no_mangle]` attr if present as well; if it is retained, it will still
        // generate an export.
        function
            .attrs
            .retain(|attr| !attr.path().is_ident("no_mangle"));

        let export = parsed_function.wrap_export(function, export_name_attr);
        (Some(parsed_function.declare(None)), Some(export))
    } else {
        (None, None)
    };

    quote! {
        #function
        #export
        #declaration
    }
}

struct Imports {
    module_name: String,
    functions: Vec<(Function, TokenStream)>,
}

impl Imports {
    fn new(module: &mut ItemForeignMod, attrs: &ExternrefAttrs) -> Result<Self, SynError> {
        const NO_ATTR_MSG: &str = "#[link(wasm_import_module = \"..\")] must be specified \
            on the foreign module";

        check_abi("foreign module", module.abi.name.as_ref(), &module.abi)?;

        let link_attr = module
            .attrs
            .iter()
            .find(|attr| attr.path().is_ident("link"));
        let Some(link_attr) = link_attr else {
            return Err(SynError::new_spanned(module, NO_ATTR_MSG));
        };

        let module_name = if matches!(link_attr.meta, Meta::List(_)) {
            let nested =
                link_attr.parse_args_with(Punctuated::<Meta, Token![,]>::parse_terminated)?;
            nested
                .into_iter()
                .find_map(|nested_meta| match nested_meta {
                    Meta::NameValue(nv) if nv.path.is_ident("wasm_import_module") => Some(nv.value),
                    _ => None,
                })
        } else {
            let msg =
                "Unexpected contents of `#[link(..)]` attr (expected a list of name-value pairs)";
            return Err(SynError::new_spanned(link_attr, msg));
        };

        let module_name =
            module_name.ok_or_else(|| SynError::new_spanned(link_attr, NO_ATTR_MSG))?;
        let module_name = if let Expr::Lit(ExprLit {
            lit: Lit::Str(str), ..
        }) = module_name
        {
            str.value()
        } else {
            let msg = "Unexpected WASM module name format (expected a string)";
            return Err(SynError::new(module_name.span(), msg));
        };

        let cr = attrs.crate_path();
        let mut functions = Vec::with_capacity(module.items.len());
        for item in &mut module.items {
            if let ForeignItem::Fn(fn_item) = item {
                let link_name = attr_expr(&fn_item.attrs, "link_name")?;
                let has_link_name = link_name.is_some();
                let function = Function::from_sig(&fn_item.sig, link_name, attrs);
                if !function.needs_declaring() {
                    continue;
                }

                let vis = mem::replace(&mut fn_item.vis, Visibility::Inherited);
                let (wrapper, new_ident) = function.wrap_import(&vis, fn_item.sig.clone());
                if !has_link_name {
                    // Add `#[link_name = ".."]` since the function is renamed.
                    let name = fn_item.sig.ident.to_string();
                    fn_item.attrs.push(syn::parse_quote!(#[link_name = #name]));
                }
                fn_item.sig.ident = new_ident;

                // Change function signature to use `usize`s in place of `Resource`s.
                for (i, arg) in fn_item.sig.inputs.iter_mut().enumerate() {
                    if function.resource_args.contains_key(&i) {
                        if let FnArg::Typed(typed_arg) = arg {
                            typed_arg.ty = Box::new(syn::parse_quote!(#cr::ExternRef));
                        }
                    }
                }
                if matches!(function.return_type, ReturnType::Resource(_)) {
                    fn_item.sig.output = syn::parse_quote!(-> #cr::ExternRef);
                }

                functions.push((function, wrapper));
            }
        }

        Ok(Self {
            module_name,
            functions,
        })
    }

    fn declarations(&self) -> impl ToTokens {
        let function_declarations = self
            .functions
            .iter()
            .map(|(function, _)| function.declare(Some(&self.module_name)));
        quote!(#(#function_declarations)*)
    }

    fn wrappers(&self) -> impl ToTokens {
        let wrappers = self.functions.iter().map(|(_, wrapper)| wrapper);
        quote!(#(#wrappers)*)
    }
}

pub(crate) fn for_foreign_module(
    module: &mut ItemForeignMod,
    attrs: &ExternrefAttrs,
) -> TokenStream {
    let parsed_module = match Imports::new(module, attrs) {
        Ok(module) => module,
        Err(err) => return err.into_compile_error(),
    };
    let declarations = parsed_module.declarations();
    let wrappers = parsed_module.wrappers();
    quote! {
        #module
        #declarations
        #wrappers
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn declaring_signature_for_export() {
        let export_fn: ItemFn = syn::parse_quote! {
            pub extern "C" fn test_export(
                sender: &mut Resource<Sender>,
                buffer: Resource<Buffer>,
                some_ptr: *const u8,
            ) {
                // does nothing
            }
        };
        let parsed = Function::new(&export_fn, &ExternrefAttrs::default()).unwrap();
        assert!(parsed.needs_declaring());

        let declaration = parsed.declare(None);
        let declaration: syn::Item = syn::parse_quote!(#declaration);
        let expected: syn::Item = syn::parse_quote! {
            externref::declare_function!(externref::Function {
                kind: externref::FunctionKind::Export,
                name: "test_export",
                externrefs: externref::BitSlice::builder::<1usize>(3usize)
                    .with_set_bit(0usize)
                    .with_set_bit(1usize)
                    .build(),
            });
        };
        assert_eq!(declaration, expected, "{}", quote!(#declaration));
    }

    #[test]
    fn transforming_export() {
        let export_fn: ItemFn = syn::parse_quote! {
            pub extern "C" fn test_export(
                sender: &mut Resource<Sender>,
                buffer: Option<Resource<Buffer>>,
                some_ptr: *const u8,
            ) {
                // does nothing
            }
        };
        let parsed = Function::new(&export_fn, &ExternrefAttrs::default()).unwrap();
        assert_eq!(parsed.resource_args.len(), 2);
        assert_eq!(parsed.resource_args[&0], SimpleResourceKind::MutRef.into());
        assert_eq!(
            parsed.resource_args[&1],
            ResourceKind::Option(SimpleResourceKind::Owned)
        );
        assert_eq!(parsed.return_type, ReturnType::Default);

        let wrapper = parsed.wrap_export(&export_fn, None);
        let wrapper: syn::Item = syn::parse_quote!(#wrapper);
        let expected: syn::Item = syn::parse_quote! {
            const _: () = {
                #[export_name = "test_export"]
                unsafe extern "C" fn __externref_export(
                    __arg0: externref::ExternRef,
                    __arg1: externref::ExternRef,
                    __arg2: *const u8,
                ) {
                    test_export(
                        externref::Resource::new(__arg0)
                            .as_mut()
                            .expect("null reference passed from host"),
                        externref::Resource::new(__arg1),
                        __arg2,
                    );
                }
            };
        };
        assert_eq!(wrapper, expected, "{}", quote!(#wrapper));
    }

    #[test]
    fn wrapper_for_import() {
        let sig: Signature = syn::parse_quote! {
            fn send_message(
                sender: &Resource<Sender>,
                message_ptr: *const u8,
                message_len: usize,
            ) -> Resource<Bytes>
        };
        let parsed = Function::from_sig(&sig, None, &ExternrefAttrs::default());

        let (wrapper, ident) = parsed.wrap_import(&Visibility::Inherited, sig);
        assert_eq!(ident, "__externref_send_message");

        let wrapper: ItemFn = syn::parse_quote!(#wrapper);
        let expected: ItemFn = syn::parse_quote! {
            #[inline(never)]
            unsafe fn send_message(
                __arg0: &Resource<Sender>,
                __arg1: *const u8,
                __arg2: usize,
            ) -> Resource<Bytes> {
                unsafe { externref::ExternRef::guard(); }
                let __output = __externref_send_message(
                    externref::Resource::raw(core::option::Option::Some(__arg0)),
                    __arg1,
                    __arg2,
                );
                externref::Resource::new(__output).expect("null reference passed from host")
            }
        };
        assert_eq!(wrapper, expected, "{}", quote!(#wrapper));
    }

    #[test]
    fn foreign_mod_transformation() {
        let mut foreign_mod: ItemForeignMod = syn::parse_quote! {
            #[link(wasm_import_module = "test")]
            extern "C" {
                fn send_message(
                    sender: &Resource<Sender>,
                    message_ptr: *const u8,
                    message_len: usize,
                ) -> Resource<Bytes>;
            }
        };
        Imports::new(&mut foreign_mod, &ExternrefAttrs::default()).unwrap();

        let expected: ItemForeignMod = syn::parse_quote! {
            #[link(wasm_import_module = "test")]
            extern "C" {
                #[link_name = "send_message"]
                fn __externref_send_message(
                    sender: externref::ExternRef,
                    message_ptr: *const u8,
                    message_len: usize,
                ) -> externref::ExternRef;
            }
        };
        assert_eq!(foreign_mod, expected, "{}", quote!(#foreign_mod));
    }
}
