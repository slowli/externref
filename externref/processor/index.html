<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="WASM module processor for `externref`s."><title>externref::processor - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-dd39b87e5fcfba68.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="externref" data-themes="" data-resource-suffix="" data-rustdoc-version="1.81.0-nightly (ed7e35f34 2024-07-06)" data-channel="nightly" data-search-js="search-d234aafac6c221dd.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../externref/index.html">externref</a><span class="version">0.3.0-beta.1</span></h2></div><h2 class="location"><a href="#">Module processor</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></section><h2><a href="../index.html">In crate externref</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">externref</a>::<wbr><a class="mod" href="#">processor</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/externref/processor/mod.rs.html#1-145">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><span class="item-info"><div class="stab portability">Available on <strong>crate feature <code>processor</code></strong> only.</div></span><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>WASM module processor for <code>externref</code>s.</p>
<p>WASM modules that use the <code>externref</code> crate need to be processed in order
to use <code>externref</code> function args / return types for imported or exported functions that
originally used <a href="../struct.Resource.html" title="struct externref::Resource"><code>Resource</code></a>s. This module encapsulates processing logic.</p>
<p>More precisely, the processor performs the following steps:</p>
<ul>
<li>Parse the custom section with <a href="../struct.Function.html" title="struct externref::Function"><code>Function</code></a> declarations and remove this section
from the module.</li>
<li>Replace imported functions from a surrogate module for handling <code>externref</code>s with
local functions.</li>
<li>Patch signatures and implementations of imported / exported functions so that they
use <code>externref</code>s where appropriate.</li>
<li>Add an initially empty, unconstrained table with <code>externref</code> elements and optionally
export it from the module. The host can use the table to inspect currently used references
(e.g., to save / restore WASM instance state).</li>
</ul>
<p>See <a href="..">crate-level docs</a> for more insights on WASM module setup and processing.</p>
<h2 id="on-processing-order"><a class="doc-anchor" href="#on-processing-order">§</a>On processing order</h2>
<p>⚠ <strong>Important.</strong> The <a href="struct.Processor.html" title="struct externref::processor::Processor"><code>Processor</code></a> should run <em>before</em> WASM optimization tools such as <code>wasm-opt</code>.
These tools may inline <code>externref</code>-operating functions, which can lead to the processor
producing invalid WASM bytecode (roughly speaking, excessively replacing <code>i32</code>s
with <code>externref</code>s). Such inlining can usually be detected by the processor, in which case
it will return <a href="enum.Error.html#variant.IncorrectGuard" title="variant externref::processor::Error::IncorrectGuard"><code>Error::IncorrectGuard</code></a> or <a href="enum.Error.html#variant.UnexpectedCall" title="variant externref::processor::Error::UnexpectedCall"><code>Error::UnexpectedCall</code></a>
from <a href="struct.Processor.html#method.process" title="method externref::processor::Processor::process"><code>process()</code></a>.</p>
<p>Optimizing WASM after the processor has an additional advantage in that it can
optimize the changes produced by it (optimization is hard, and is best left
to the dedicated tools).</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>externref::processor::Processor;

<span class="kw">let </span>module: Vec&lt;u8&gt; = <span class="comment">// WASM module, e.g., loaded from the file system
</span><span class="kw">let </span>processed: Vec&lt;u8&gt; = Processor::default()
    <span class="comment">// Set a hook to be called when a reference is dropped
    </span>.set_drop_fn(<span class="string">"test"</span>, <span class="string">"drop_ref"</span>)
    .process_bytes(<span class="kw-2">&amp;</span>module)<span class="question-mark">?</span>;
<span class="comment">// Store or use the processed module...</span></code></pre></div>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Processor.html" title="struct externref::processor::Processor">Processor</a></div><div class="desc docblock-short">WASM module processor encapsulating processing options.</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.Error.html" title="enum externref::processor::Error">Error</a></div><div class="desc docblock-short">Errors that can occur when <a href="struct.Processor.html#method.process" title="method externref::processor::Processor::process">processing</a> a WASM module.</div></li><li><div class="item-name"><a class="enum" href="enum.Location.html" title="enum externref::processor::Location">Location</a></div><div class="desc docblock-short">Location of a <code>Resource</code>: a function argument or a return type.</div></li></ul></section></div></main></body></html>