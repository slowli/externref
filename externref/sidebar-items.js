window.SIDEBAR_ITEMS = {"attr":["externref"],"enum":["FunctionKind","ReadErrorKind"],"mod":["processor"],"struct":["BitSlice","Function","ReadError","Resource"]};