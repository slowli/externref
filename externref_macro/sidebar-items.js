window.SIDEBAR_ITEMS = {"attr":["externref"]};