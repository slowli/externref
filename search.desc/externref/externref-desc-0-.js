searchState.loadedDescShard("externref", 0, "Low-cost reference type shims for WASM modules.\nSlice of bits. This type is used to mark <code>Resource</code> args in …\nName of a custom section in WASM modules where <code>Function</code> …\nFunction exported from a WASM module.\nInformation about a function with <code>Resource</code> args or return …\nKind of a function with <code>Resource</code> args or return type.\nFunction imported to a WASM module from the module with …\nErrors that can occur when reading declarations of …\nKind of a <code>ReadError</code>.\nHost resource exposed to WASM.\nUnexpected end of the input.\nError parsing\nReturns the number of bits in this slice.\nReturns the number of set bits in this slice.\nDowncasts this generic resource to a specific type.\nPrepares imported functions or an exported function with …\nBit slice marking <code>Resource</code> args / return type.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nChecks if a bit with the specified 0-based index is set.\nKind of this function.\nName of this function.\nWASM module processor for <code>externref</code>s.\nReads function information from a WASM custom section. …\nIterates over the indexes of set bits in this slice.\nUpcasts this resource to a generic resource.\nUpcasts a reference to this resource to a generic resource …\nArgument with the specified zero-based index.\nErrors that can occur when processing a WASM module.\nIncorrectly placed <code>externref</code> guard. This is caused by …\nLocation of a <code>Resource</code>: a function argument or a return …\nMissing exported function with the enclosed name.\nWASM module processor encapsulating processing options.\nError reading the custom section with function …\nReturn type with the specified zero-based index.\nImported or exported function has unexpected arity.\nUnexpected call to a function returning <code>externref</code>. Such …\nUnexpected type of an export (expected a function).\nUnexpected type of an import (expected a function).\nArgument or return type of a function has unexpected type.\nError parsing the WASM module.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nProcesses the provided <code>module</code>.\nProcesses the provided WASM module <code>bytes</code>. This is a …\nSets a function to notify the host about dropped <code>externref</code>…\nSets the name of the exported <code>externref</code>s table where refs …\nWASM bytecode offset of the offending guard.\nWASM bytecode offset of the offending call.\nExpected arity of the function.\nName of the function with an incorrectly placed guard.\nName of the function containing an unexpected call.\nLocation of an argument / return type in the function.\nName of the module.\nName of the module; <code>None</code> for exported functions.\nName of the module; <code>None</code> for exported functions.\nName of the function.\nName of the function.\nName of the function.\nActual arity of the function.\nActual type of the function (the expected type is always …")