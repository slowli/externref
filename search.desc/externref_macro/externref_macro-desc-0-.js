searchState.loadedDescShard("externref_macro", 0, "Procedural macro for <code>externref</code>.\nPrepares imported functions or an exported function with â€¦")