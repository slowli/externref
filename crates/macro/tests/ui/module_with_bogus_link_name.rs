use externref_macro::externref;

#[externref]
#[link(wasm_import_module = "test")]
extern "C" {
    #[link_name("huh")]
    pub fn unused(ptr: *const u8, len: usize);
}

#[externref]
#[link(wasm_import_module = "test")]
extern "C" {
    #[link_name = 3]
    pub fn other(ptr: *const u8, len: usize);
}

fn main() {}
