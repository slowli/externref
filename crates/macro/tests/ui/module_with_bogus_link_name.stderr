error: expected `=`
 --> tests/ui/module_with_bogus_link_name.rs:6:16
  |
6 |     #[link_name("huh")]
  |                ^

error: malformed `link_name` attribute input
  --> tests/ui/module_with_bogus_link_name.rs:13:5
   |
13 |     #[link_name = 3]
   |     ^^^^^^^^^^^^^^^^ help: must be of the form: `#[link_name = "name"]`
